### PR TITLE
[dart][dart-dio] Next-gen dart-dio generator

### PR DESCRIPTION
**Pull request by kuhnroyal**
### WIP!
This is work in progress but examples compile and tests run locally. This needs an updated Dart version in CI.

### Description

This PR establishes a next-gen dart-dio generator with breaking changes around null-safety which can not be implemented in backwards-compatible ways.

**Features & changes**
* Dart 2.12.0+ requirement
* Dio 4.0.0+ requirement
* null-safety implemented
* new package layout with a single exported file
* wrapping of serialization errors in `DioError`
* named parameters for all API params
* support for different serialization libraries with built_value being the current only and default implementation
* enforce `dart format` in CI (waiting on https://github.com/dart-lang/sdk/issues/44582)

### TODO
- [x] need a solution for serializing `null` with built_value (https://github.com/google/built_value.dart/issues/912)
- [x] need to handle `unused_import` analyzer warning in polymorph model classes
- [ ] wait for timemachine to support NNBD (https://github.com/Dana-Ferguson/time_machine/issues/55)

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by kuhnroyal** [comment]
CC @agilob @josh-burton

**Comment by robrbecker** [comment]
I'm having a heck of a time getting this to build locally. I have tried to build with openjdk8 and oracle java 14.0.2.
Had to comment out the javadocs plugin from pom.xml as that was crashing with an exception.

Still getting an exception when running `./bin/generate-samples.sh bin/configs/dart-dio-next*`

```
[main] INFO  o.o.codegen.TemplateManager - writing file samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake/doc/DogAllOf.md
[main] WARN  o.o.c.languages.AbstractDartCodegen - 200_response (model name starts with number) cannot be used as model name. Renamed to Model200Response
[main] ERROR o.o.c.api.TemplatingEngineAdapter - Failed to read full template enum_inline.mustache, enum_inline.mustache
[main] ERROR o.o.c.api.TemplatingEngineAdapter - Failed to read full template enum_inline.mustache, enum_inline.mustache
Exception in thread "main" java.lang.RuntimeException: Could not generate model 'EnumArrays'
	at org.openapitools.codegen.DefaultGenerator.generateModels(DefaultGenerator.java:544)
	at org.openapitools.codegen.DefaultGenerator.generate(DefaultGenerator.java:875)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:440)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
Caused by: org.openapitools.codegen.templating.TemplateNotFoundException: enum_inline
	at org.openapitools.codegen.templating.MustacheEngineAdapter.findTemplate(MustacheEngineAdapter.java:79)
	at org.openapitools.codegen.templating.MustacheEngineAdapter.lambda$compileTemplate$0(MustacheEngineAdapter.java:61)
	at com.samskivert.mustache.Mustache$IncludedTemplateSegment.execute(Mustache.java:756)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$InvertedSegment.execute(Mustache.java:910)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$SectionSegment.execute(Mustache.java:870)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$SectionSegment.execute(Mustache.java:866)
	at com.samskivert.mustache.Template.executeSegs(Template.java:157)
	at com.samskivert.mustache.Mustache$IncludedTemplateSegment.execute(Mustache.java:774)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$SectionSegment.execute(Mustache.java:881)
	at com.samskivert.mustache.Template.executeSegs(Template.java:157)
	at com.samskivert.mustache.Mustache$IncludedTemplateSegment.execute(Mustache.java:774)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$InvertedSegment.execute(Mustache.java:910)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$SectionSegment.execute(Mustache.java:881)
	at com.samskivert.mustache.Mustache$BlockSegment.executeSegs(Mustache.java:845)
	at com.samskivert.mustache.Mustache$SectionSegment.execute(Mustache.java:866)
	at com.samskivert.mustache.Template.executeSegs(Template.java:157)
	at com.samskivert.mustache.Template.execute(Template.java:134)
	at com.samskivert.mustache.Template.execute(Template.java:125)
	at org.openapitools.codegen.templating.MustacheEngineAdapter.compileTemplate(MustacheEngineAdapter.java:65)
	at org.openapitools.codegen.TemplateManager.write(TemplateManager.java:163)
	at org.openapitools.codegen.DefaultGenerator.processTemplateToFile(DefaultGenerator.java:1017)
	at org.openapitools.codegen.DefaultGenerator.generateModel(DefaultGenerator.java:382)
	at org.openapitools.codegen.DefaultGenerator.generateModels(DefaultGenerator.java:535)
```

**Comment by robrbecker** [comment]
I think I got it running enough to generate my project by running something like
`java -jar openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate ... <params>`

**Comment by kuhnroyal** [comment]
Hmm, the error looks like I might have messed something up with the templates but the example generates, not sure.

> I think I got it running enough to generate my project by running something like
> `java -jar openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate ... <params>`

That is the way I am using it outside of the build in examples.

**Comment by vishna** [comment]
Just stumbled upon this, looks promising - good job 👍 I was wondering if it would be easier to use `freezed` instead of `build_value`?

**Comment by kuhnroyal** [comment]
Good question and one I have evaluated before. `freezed` does not support the level of enums that are required for OpenAPI - see https://github.com/rrousselGit/freezed/issues/74

But this new version of the generator will allow to add different serialization libraries in the future. You are welcome to start working on a freezed implementation and see where that leads you.

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team as the change covers the codegen operation class.

**Comment by kuhnroyal** [comment]
> cc @OpenAPITools/generator-core-team as the change covers the codegen operation class.

Yes, this is a minor change to simplify templates as it is hard to do `OR` checks there.

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/auth/basic_auth.mustache [hunk 1]
```diff
@@ -1,33 +1,34 @@
...
+  final String username;
+  final String password;
 
-    const BasicAuthInfo(this.username, this.password);
+  const BasicAuthInfo(this.username, this.password);
 }
 
 class BasicAuthInterceptor extends AuthInterceptor {
-    Map<String, BasicAuthInfo> authInfo = {};
+  Map<String, BasicAuthInfo> authInfo = {};
```
Comment (agilob) [inline modules/openapi-generator/src/main/resources/dart/libraries/dio/auth/basic_auth.mustache:15]: final for consistency?

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/auth/oauth.mustache [hunk 2]
```diff
@@ -1,22 +1,23 @@
 {{>header}}
-import 'dart:async';
-
 import 'package:dio/dio.dart';
 import 'package:{{pubName}}/src/auth/auth.dart';
 
 class OAuthInterceptor extends AuthInterceptor {
-    Map<String, String> tokens = {};
+  Map<String, String> tokens = {};
```
Comment (agilob) [inline modules/openapi-generator/src/main/resources/dart/libraries/dio/auth/oauth.mustache:6]: final for consistency?

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/auth/basic_auth.mustache (hunk 1 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/auth/basic_auth.mustache:15]: Sure, this was just a copy/paste job from dart-dio.

**Comment by jannikrh** [comment]
If the additional library timemachine is not used, the Date format of the API definition cannot be used. A complete DateTime string is always generated. Is it possible to distinguish between Date and DateTime like in the "normal" Dart generator?

As an example the following formatter is used in the Dart generator: `final _dateFormatter = DateFormat('yyyy-MM-dd');`

**Comment by agilob** [comment]
>If the additional library timemachine is not used, the Date format cannot be used. A complete DateTime string is always generated. Is it possible to distinguish between Date and DateTime like in the "normal" Dart generator?

Could you open an issue for that and call me and @kuhnroyal  in that?

**Comment by kuhnroyal** [comment]
Don't think this is possible unfortunately, but I have not tried. built_value can only have one serializer per class and date/datetime is the same class `DateTime` in plain Dart.

**Comment by jannikrh** [comment]
I will open an issue later today and tag you. The implementation might be possible by extending Iso8601DateTimeSerializer in combination with FullType parameters. The DateTime serializer would be called in each case. Once as now and once with the parameter date or similar.

**Comment by agilob** [comment]
lets not comment here again, we will lose context of the problem. call us to the issue

**Comment by TimWhiting** [comment]
@kuhnroyal 
I'd love to have freezed support. I'm not understanding the relation of https://github.com/rrousselGit/freezed/issues/74 to enum support in OpenAPI. Is it a problem with modeling enums or serializing? Serializing should be taken care of by JsonSerializable shouldn't it? Can you elaborate more on the issues with freezed support?

**Comment by Abhilash-Chandran** [comment]
@TimWhiting I already have a version that uses freezed. I just need to clean it up a bit and raise a PR which probably @kuhnroyal could validate. I am using normal enum and seems to work well. Will create a PR today... 😄

**Comment by TimWhiting** [comment]
@Abhilash-Chandran I'm happy to test it on my apis to see if it makes sense. Can you push what you have so far to a branch?

**Comment by dastein1** [comment]
@Abhilash-Chandran I'd be interested in this as well.

**Comment by TimWhiting** [comment]
@dastein1 @Abhilash-Chandran 
Abhilash, I've taken your branch that you gave me the link to, and squashed your commits, and rebased on master. It has a much cleaner PR / diff now. However, I think I broke something in the rebase. Anyways, for those who are interested, Abhilash's branch does work even if it is a bit out of date / sync with the rest of the repo:

Here is a link to @Abhilash-Chandran's branch:
https://github.com/bahag-chandrana/openapi-generator
Here is the link to the rebased changes:
https://github.com/TimWhiting/openapi-generator/tree/freezed_2

If anyone wants to try to get the rebased branch working & a PR opened, that would probably be a good place to start. I'll see what time I have, but I can't make any promises. I was hoping to see `allOf` and `oneOf` support with inheritance & freezed unions, but unfortunately that looks like it will be more work.

Here are the instructions I got from @Abhilash-Chandran to run his branch: 
You can find the jar inside modules/openapi-generator-cli/target/openapi-generator-cli.jar
Once its merged with main it should be available via the usual cli commands.
`
java -jar ../openapi-generator-cli.jar generate \
      -i api_spec.yaml \
      -g dart-dio-next \
      -c generator_config.yaml 
`

with the generator_config.yaml containing the additional property to enable freezed.
```yaml
additionalProperties:
  serializationLibrary: freezed
```

**Comment by Abhilash-Chandran** [comment]
I saw some simiallar issues while building after rebasing. I will try to rework on this from the fresh branch. But no promises yet. Especially the way the templates are referred to has changed since I began working on this. So my idea is to port my changes directly on a fresh fork and hence avoid a host of other errors.

**Comment by TimWhiting** [comment]
@Abhilash-Chandran I did end up copying the json_serializable/api folder to the freezed folder. That seemed to be the only major change needed to get it to compile & generate valid dart files. However, I have not tried using the generated api yet. The changes are on my branch.

**Pull request merged**